### ISSUE TITLE
Why do you carry true_boxes during the training?

### ISSUE THREAD

**Issue by bshao001** [experiencor/basic-yolo-keras#42]
I don't understand why you carry true_boxes in the input of the model. The ground truth is already in y_true. I saw this effort impact the yolo model (backend), and the BatchGenerator, and then the loss function (custom_loss). I assume the intention is to use this information in the loss function. However, why cannot just extract that from y_true parameter instead of using self.true_boxes (I assume this self.true_boxes is constantly receiving the ground truth bounding box information during the training)?

Thank you very much for the time and help!

**Comment by experiencor** [comment]
@bshao001 You are right. We can extract true_boxes directly from the y_true parameter. But this procedure is way too complicated in Tensorflow. I gave up halfway, and simply feed the redundant parameter true_boxes for simplicity. But you may try.

**Comment by bshao001** [comment]
@experiencor Thank you very much for the explanation. 

Instead of opening a new issue, I would like to ask another question here:

I have not carefully studied the original paper, however, based on the recent lecture from Andrew Ng from his Coursera.org, line 178 in preprocessing.py should be something like this:

    box = [center_x - grid_x, center_y - grid_y, center_w, center_h]

If you do implement the BatchGenerator this way, I assume in the custom_loss function in frontend.py, you don't need to create cell_grid variable and add it to pred_box_xy. 

However, I am really confused here. Since you do feed the training data with the base of the grid value included (for example, 3.25, 4.12 instead of 0.25, 0.12), the prediction should include the base (in the example, 3 and 4).

It is also weird that you apply tf.sigmoid to the y_pred for the box_xy value which turns regression type values into a classification type values. Is that the reason you then can/need add cell_grid? If so, it looks wrong here.

Another thing I don't understand is line 108 in frontend.py. Applying tf.exp to it sounds so strange.

Thank you again for your time and help!

**Comment by experiencor** [comment]
1. I don't do this [center_x - grid_x, center_y - grid_y, center_w, center_h] but do this [center_x, center_y, center_w, center_h]. The reason is that I need pred_box_xy = tf.sigmoid(y_pred[..., :2]) + cell_grid to compute the overlap between each predicted boxes and each true_boxes (in the "confidence mask" computation. This computation is not constrained to the cells.

2. Yes, in line "pred_box_xy = tf.sigmoid(y_pred[..., :2]) + cell_grid", I use cell_grid to shift the predictions.

3. The position of the box within a cell not including the offset grid varies from 0 to 1. (0,0) for top left corner of the cell. (1,1) for bottom right. So we use tf.sigmoid to scale the prediction x and y to the interval [0,1]. We regress the relative position of a box in this interval.

4. This line simply scales the w and h of 5 bounding boxes using the sizes of the anchors. It's very simple but the Tensorflow operation really obfuscates its meaning.

**Comment by bshao001** [comment]
@experiencor Although I am not convinced, your answer is quite clear and I really appreciate that. Thank you very much.
